File: src/javascript/algorithms/array/containsDuplicate.js
// Early exist:- TC: O(n), SC: O(1)
function containsDuplicate(nums) {
    let noDupsSet = new Set();
    for(const num of nums) {
        if(noDupsSet.has(num)) {
            return true;
        }
        noDupsSet.add(num);
    }
    return false;
}

//Use Set size:- TC: O(n), SC: O(1)
function containsDuplicateUsingSize(nums) {
    return new Set(nums).size !== nums.length;
}

console.log("-----Has duplicates----");
let nums1 = [8, 6, 4, 2, 6];
console.log(containsDuplicate(nums1));
console.log(containsDuplicateUsingSize(nums1));
console.log("-----No duplicates----");
let nums2 = [1, 3, 5, 7, 9];
console.log(containsDuplicate(nums2));
console.log(containsDuplicateUsingSize(nums2));

File: src/java1/algorithms/array/containsDuplicate/ContainsDuplicate.java
package containsDuplicate;
import java.util.*;

class ContainsDuplicate {
    private static boolean containsDuplicate(int[] nums) {
        HashSet<Integer> set = new HashSet<>();

        for (int num : nums) {
            if (set.contains(num)) {
                return true;
            }
            set.add(num);
        }
        return false;
    }

    private static boolean containsDuplicateUsingMap(int[] nums) {
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int num : nums) {
            if (map.containsKey(num)) {
                return true;
            }
            map.put(num, 1);
        }
        return false;
    }

    public static void main(String[] args) {
        int[] nums1 = new int[] { 8, 6, 4, 2, 6 };
        int[] nums2 = new int[] { 1, 3, 5, 7, 9 };

        System.out.println(containsDuplicate(nums1));
        System.out.println(containsDuplicate(nums2));

        System.out.println(containsDuplicateUsingMap(nums1));
        System.out.println(containsDuplicateUsingMap(nums2));
    }
}